### ISSUE TITLE
Need API exposure of flow status

### ISSUE THREAD

**Issue by closer8888** [node-red/node-red#698]
I have created a wrapper for Node-Red that can spawn many instances running on consecutive ports and proxied through the parent process. Each child instance can be stopped and started without being destroyed. What I need is a way to see if a particular instance is flowing or not flowing. Something like this:
If you take instruction from website Documentation for "Embedding into an existing app" the Red object  has functions start() and stop(). It needs a status() function. This function could expose the property from Red.server.redNodes.flows.activeFlow.started

**Comment by kevinroisin** [comment]
I would support an optional callback being added to `RED.start()` which would get triggered when the flows are started.

**Comment by levidobson** [comment]
Indicating that flows have been started is good, however we also need a way to check whether all flows have finished flowing, and it is safe to stop / reload flows. Optionally, it could also be useful if the stop call would wait for all flows to finish, before actually stopping and returning / resolving.

**Comment by knolleary** [comment]
@levidobson it isn't possible to know that the flows have 'finished' as the runtime doesn't know what finished means for your application.

For example, you may have a flow that halts half-way through based on some condition. Or a node may be making a long-running async call to an external service. The runtime doesn't know if a node is sat idle, or sat waiting for an async response.

**Comment by levidobson** [comment]
@knolleary yes, I understand the problem. However this means, that currently a graceful shutdown scenario is impossible. Also, if you want to dynamically load flows based on some input parameters (which is what I'm trying to do), you cannot make sure that when you load the new flow, the current flow has already stopped running.

I would say at first glance that extending the node api to support a "started" and a "completed" flag would do the trick, then the engine would have all the necessary information about a flow to determine if it has already started and completed.

For graceful shutdown currently the best option is to restrict input channels and wait for a pre-defined time period before shutting down / restarting the process. Or is there a better way?

**Comment by dceejay** [comment]
Even if you know a flow has started... but not completed... what are you going to do ? you still have to wait and restart... so you're no better off ?
When you re-deploy each node's close() method gets called and well behaved ones then call done() when complete.

**Comment by knolleary** [comment]
This isn't about individual nodes closing cleanly - we do that well today. What we don't do is shutdown with any consideration of the flows as a whole. As I said previously, we currently don't know anything about the intended operation of the application. If a message is in the middle of the flow when we shutdown, it will get dropped on the floor. That is not ideal if you want any reliability over redeploys/restarts.

It's a hard thing to solve from where we are today and would require some additions to how nodes define themselves, so that, for example:
 - Input nodes are formally identified as 'message generators' so they can be quiesced first to stop new work coming into a flow
 - Have some mechanism for a node to indicate it is still processing a message (which would also benefit the metric/monitoring of flows)
 - ensure appropriate fallback behaviours for existing nodes - ultimately timing out to ensure things don't hang indefinitely like they do today if a node doesn't call it's `done` function
 - and other things that we haven't thought of yet from this short conversation

Whilst it isn't something we can address in the short-term, I don't think it is something we can ignore; it _will_ be an inhibitor to some users adopting Node-RED.
